Searching

brewingcode opened the pull request:
This is modeled after vim's search behavior. Using this as an example:

```bash
$ fx <(curl -s https://next.json-generator.com/api/json/get/VkCpIMn1U)
```

1. Hit `/` to start/resume searching
2. Enter `strawberry` as a plaintext search, and then hit `<enter>`
3. You'll see `1 of 4 found: [0].favoriteFruit`, and the JSON will expand to that path and put the cursor on the proper line: at that point, you can move around as normal, with the addition of `n` to go to the next search hit, and `p` to go back to the previous one.
4. You can also use a regex: hit `/`, and the `C-u` to erase the previous search and enter `/t(ue|hur)sday/i` (note the regex syntax), and you'll see two hits
5. Searching also includes the keys, ie `eyeColor` will find the 9 "eyeColor" keys in the example JSON

The majority of this search functionality is in its own [search module](https://github.com/brewingcode/fx/blob/searching/search.js), the only changes to `fx.js` are:

1. Adding an event listener `box.on('focus', ...)` which is how the search module tells the main screen to update

2. Modifying the `render` method to optionally take a `path` argument, which the `focus` handler uses to tell `render()` which line to put the cursor on

I also grouped some of the smaller functions lying around into [helpers.js](https://github.com/brewingcode/fx/blob/searching/helpers.js): `walk` (which I added a few optional arguments to), `reduce`, and the new `log` function are all in there. This was mostly because I needed `walk` in  [search.js](https://github.com/brewingcode/fx/blob/searching/search.js). Also, that `log()` function doesn't do anything unless you define a `log: "/tmp/some-file-name.log"` entry in `config.js`.

brewingcode commented:
Hahaha, #51 was opened about the same time I started working on this branch. Didn't even notice!

Also, I just pushed 30b035b to fix the tests failing: I botched the rebase, and just had to properly copy the `reduce()` function to `helpers.js`.

antonmedv commented:
Awesome work! 🔥 
I'll review it later today.

antonmedv commented:
Found a bug: you can still press `.` and and type while search is on. But can't see what you are typing.

antonmedv commented:
Will be cool to have some sort of highlight of found part.

antonmedv commented:
Also got this error:
```
TypeError: [(...index)].find is not a function or its return value is not iterable
    at render (/Users/anton/Projects/fx/fx.js:405:37)
```

antonmedv commented:
![image](https://user-images.githubusercontent.com/141232/49992709-6e559280-ffb7-11e8-9d51-1cd4faad59b5.png)
Also I don't really like two line UI. Maybe use like `less` search?

antonmedv commented on config.js:
```diff
@@ -20,4 +20,7 @@ module.exports = {
   list:    global.FX_STYLE_LIST    || list,
+
+  // set this to a filename to send debug/tracing output there
+  // log: '/tmp/fx.log',
```
I this is not needed. fx can be debugged with `console.error` and `fx 2>err.log`

brewingcode commented:
I'll get all those points addressed, thanks @antonmedv

antonmedv commented:
@brewingcode I've implemented my own search PR: https://github.com/antonmedv/fx/pull/56
Take a look at it. Maybe we can merge our PR somehow.

brewingcode commented:
> @brewingcode I've implemented my own search PR: #56
> Take a look at it. Maybe we can merge our PR somehow.

Cool! I was able to drop-in replace `print.js` and `config.js` in my branch, and make some small modifications to get highlighting working. I have a slight preference for my own non-generator search solution, as it already allows going both forward and backward through the list of search hits.

I've also changed/fixed the rest of the points you brought up.

antonmedv commented:
Thanks. I got a few problems with interaction between `.` dot and `/` search. I'm gonna merge #56 with a few improvements from your code.

antonmedv commented:
Closing. #56 merged.
